Add device trees package for Amlogic generic project

kszaq opened the pull request:
This PR adds `KERNEL_EXTRA_DEPENDS_TARGET` variable that can be set in project options to set additional `PKG_DEPENDS_TARGET` for `linux` package.

Device trees package contains overlays that have to be compiled in kernel tree. Keeping them in a separate package allows easier maintenance and compiling only the files we want, i.e. included in the package instead of all device trees provided in kernel.

CvH commented on projects/Amlogic/bootloader/release:
```diff
@@ -0,0 +1,25 @@
+#!/bin/sh
+
+################################################################################
+#      This file is part of LibreELEC - https://LibreELEC.tv
```
`https://libreelec.tv`

CvH commented on projects/Amlogic/bootloader/release:
```diff
@@ -0,0 +1,25 @@
+
+################################################################################
+#      This file is part of LibreELEC - https://LibreELEC.tv
+#      Copyright (C) 2016 Team LibreELEC
```
`Copyright (C) 2016-present Team LibreELEC`

CvH commented on projects/Amlogic/packages/aml-s9xx-device-trees/package.mk:
```diff
@@ -0,0 +1,49 @@
+################################################################################
+#      This file is part of LibreELEC - http://www.libreelec.tv
```
`https://libreelec.tv`

CvH commented on projects/Amlogic/packages/aml-s9xx-device-trees/package.mk:
```diff
@@ -0,0 +1,49 @@
+################################################################################
+#      This file is part of LibreELEC - http://www.libreelec.tv
+#      Copyright (C) 2016 Mateusz Krzak (kszaquitto@gmail.com)
```
Idk if you happy with that or not? 
Preferred is `Copyright (C) 2016-present Team LibreELEC` but I won't beat you to remove your copyright.

or maybe

```
Copyright (C) 2018-present Team LibreELEC
Copyright (C) 2016 Mateusz Krzak (kszaquitto@gmail.com)
```
whatever you like :)

CvH commented on projects/Amlogic/packages/aml-s9xx-device-trees/package.mk:
```diff
@@ -0,0 +1,49 @@
+
+PKG_NAME="aml-s9xx-device-trees"
+PKG_VERSION="097d287"
+PKG_LICENSE="OSS"
```
`GPL` I think ?

CvH commented on projects/Amlogic/packages/aml-s9xx-device-trees/package.mk:
```diff
@@ -0,0 +1,49 @@
+PKG_NAME="aml-s9xx-device-trees"
+PKG_VERSION="097d287"
+PKG_LICENSE="OSS"
+PKG_URL="https://github.com/kszaq/s905-device-trees/archive/$PKG_VERSION.tar.gz"
```
shouldn't we fork that under our account ?
`https://github.com/LibreELEC/s905-device-trees/archive/$PKG_VERSION.tar.gz`
@chewitt @lrusak ?

chewitt replied:
I'm fine with it as-is for now, but long term we should have the project host something as our experience shows people contribute more to a project repo than a personal one.

chewitt replied:
NB: the same applies to a few other things among recent Amlogic changes, but I see them as an admin/housekeeping task to be done as we approach beta release.

adamg88 replied:
There has been a number of community contributions since the fd628-aml merge, it seems to have perked peoples interest and I've had 3 dtbs submitted to me in the last 24 hours so I would be in favour of a fork for people to contribute to.

chewitt replied:
https://github.com/LibreELEC/device-trees-amlogic is open for business .. done as a root fork; not to disrespect where things started but simply to ensure we are the obvious upstream/root target for things.

CvH commented on projects/Amlogic/packages/aml-s9xx-device-trees/package.mk:
```diff
@@ -0,0 +1,49 @@
+PKG_SOURCE_DIR="s905-device-trees-$PKG_VERSION*"
+PKG_DEPENDS_TARGET="toolchain"
+PKG_TOOLCHAIN="manual"
+PKG_IS_KERNEL_PKG="yes"
```
```
PKG_IS_KERNEL_PKG="yes"
PKG_TOOLCHAIN="manual"
```
switch lines

adamg88 commented:
Completely different implementation of how I have implemented this but I like it :) if you are including this for all Amlogic projects can the dtb.img line be removed from 'projects/Amlogic/devices/LePotato/bootloader/release' ?

CvH commented on projects/Amlogic/packages/device-trees-amlogic/package.mk:
```diff
@@ -0,0 +1,49 @@
+
+PKG_NAME="device-trees-amlogic"
+PKG_VERSION="68267ba"
+PKG_SHA256=""
```
sha256 missing :)

kszaq replied:
This is WIP, I will add SHA256 when package version is ready to go. For now PKG_VERSION only is provided for easier updates (for me) - and it works.

MilhouseVH commented:
If we're adding new project options I think we should add the new project option to all project option files, in this case it would be added to the other projects as:
```
  # additional kernel dependencies
    KERNEL_EXTRA_DEPENDS_TARGET=""
```
See `KERNEL_MAKE_EXTRACMD` and `PROJECT_CFLAGS` as existing examples.

kszaq commented:
@MilhouseVH What would be a preferred solution here? A new project option or a conditional suggested by Adam? Or perhaps there's a cleaner way to add a package that injects device trees into kernel tree?

MilhouseVH commented:
@kszaq it's up to you, really - I honestly don't know if this functionality will be of benefit to other projects in which case a conditional solution might be sufficient. If you think it might be of benefit to other projects (including those which might not yet exist) then a more "generic" solution using the project option would be the way to go.

If you take the project option route then I'm just saying that when new project options are added (which could be used by other projects in addition to the project it is being added for) then we should add the option with a suitable "do nothing" default to all projects, as it helps to "self-document" each project.

Most new developers probably only look at one project, or when bringing up a new board one that is similar to an existing project, so if options aren't included in the project they are looking at then they may never know about them.

This isn't really a hard rule as some project options may be very project specific in which case it would make no sense adding them to every project. It does mean a bit more work of course, but if the project option is the right solution then I think it should be worth it in the long run.

adamg88 commented:
@kszaq I only suggested the above as that seems to be the standard of how packages are included in other projects, either way works just as good but sticking with the original I think @milhousevh is right and it should be added to all projects or distributions/LibreELEC/options so people know it exists for possible addition to future projects.

kszaq commented:
Rebased and amended:
* add `KERNEL_EXTRA_DEPENDS_TARGET` to all projects' `options`
* amended `device-trees-amlogic` package to also parse trees already present in kernel as we also need to include some of them
* fixed including too many device trees in `dtb.img` that ships with SD card image
* added SHA256

The PR is considered complete but I welcome any further suggestions for improvements.

adamg88 commented:
@kszaq can you address the lepotato patch with this PR as in my previous comment

adamg88 commented:
Also please add `KERNEL_EXTRA_DEPENDS_TARGET` to distributions/LibreELEC/options similar to `ADDITIONAL_PACKAGES`

kszaq commented:
LePotato patch is already removed. This device is due for a room service after https://github.com/LibreELEC/LibreELEC.tv/pull/2421 is merged.

adamg88 commented:
I didn't see you had removed it, apologies for that

adamg88 commented:
If #2431 is merged before this @kszaq you may want to add $KERNEL_EXTRA_DEPENDS_TARGET to PKG_STAMP in `packages/linux/package.mk`

kszaq commented:
@adamg This is a bit of circular dependency as I'd like `device-trees-amlogic` to be rebuilt as well when `linux` changes. I think it's best to wait for #2431 and figure out a best solution then.

Meanwhile I will update package version.

adamg88 commented:
@kszaq I think the PR by kwiboo will help with this issue but yeah I agree, it's only a problem when doing lazy rebuilds anyway.

adamg88 commented on projects/Amlogic/packages/device-trees-amlogic/package.mk:
```diff
@@ -0,0 +1,62 @@
+    DTB_NAME="$(basename $f .dts).dtb"
+    DTB_LIST="$DTB_LIST $DTB_NAME"
+  done
+
```
Since your last PR to device-trees package some code is needed to add *.dtsi, possibly ...
```
  # Copy all device tree includes to kernel source folder
  for f in $PKG_BUILD/*.dtsi; do
    cp -f $f arch/$TARGET_KERNEL_ARCH/boot/dts/amlogic/
  done
```

kszaq replied:
Look 2 lines above; `*.dts*`

adamg88 replied:
Aha :s it didn't copy the files for me when I was testing earlier, lepotato.dtsi missing

adamg88 replied:
Sorry I didn't realize you had updated the commits in this PR

adamg88 commented on projects/Amlogic/packages/device-trees-amlogic/package.mk:
```diff
@@ -0,0 +1,62 @@
+
+  # Device trees already present in kernel tree we want to include
+  EXTRA_TREES=(gxbb_p201.dtb gxl_p212_1g.dtb gxl_p212_2g.dtb \
+               gxm_q200_2g.dtb gxm_q201_1g.dtb gxm_q201_2g.dtb)
```
missing gxl_p281_1g

kszaq commented:
If there are no objections, this should be good to go.

kszaq commented on projects/Amlogic/bootloader/update.sh:
```diff
@@ -99,6 +99,11 @@ for arg in $(cat /proc/cmdline); do
 
+if [ -d $BOOT_ROOT/device_trees ]; then
+  rm $BOOT_ROOT/device_trees/*.dtb
+  cp -p $SYSTEM_ROOT/usr/share/bootloader/*.dtb $BOOT_ROOT/device_trees/
```
Correct, I will add it. Thanks.

kszaq commented on projects/Amlogic/options:
```diff
@@ -54,14 +54,16 @@
     KERNEL_MAKE_EXTRACMD=""
 
+  # Additional kernel dependencies
+    KERNEL_EXTRA_DEPENDS_TARGET="device-trees-amlogic"
```
This package is also required for LePotato. It does no harm for other devices but I can move it if required.

Raybuntu replied:
OK you can leave it if it doesn't break updates on WH, WP2 I think.

kszaq replied:
Sorry, already changed. ;)

kszaq commented on projects/Amlogic/packages/device-trees-amlogic/package.mk:
```diff
@@ -0,0 +1,63 @@
+  pushd arch/$TARGET_KERNEL_ARCH/boot/dts/amlogic > /dev/null
+  for f in ${EXTRA_TREES[@]}; do
+    LE_DT_ID="${f/.dtb/}"
+    echo -e "/ {\n\tle-dt-id = \"$LE_DT_ID\";\n};" >> "${f/.dtb/.dts}"
```
The reason is that I'm too lazy to push 2 PRs (kernel + this repo) in case there is a need to add dtb from kernel tree.

adamg88 commented on projects/Amlogic/packages/device-trees-amlogic/package.mk:
```diff
@@ -0,0 +1,81 @@
+  # Device trees already present in kernel tree we want to include
+  EXTRA_TREES=(gxbb_p201 gxl_p212_1g gxl_p212_2g gxm_q200_2g gxm_q201_1g gxm_q201_2g gxl_p281_1g )
+
+  # Add "le-dtb-id" and add trees to the list
```
Please look at the following PR so this portion of code can be removed https://github.com/LibreELEC/linux-amlogic/pull/97
